shared/typeinference/codeql/typeinference/internal/TypeInference.qll

```diff
@@ -0,0 +1,776 @@
+         * have:
+         *
+         * `baseMention` | `path`    | `t`
+         * ------------- | --------- | ---
+         * `Mid<C<T4>>`  | `"0"`     | ``C`1``
+         * `Mid<C<T4>>`  | `"0.1"`   | `int`
+         * `Base<C<T3>>` | `"0"`     | ``C`1``
+         * `Base<C<T3>>` | `"0.0"`   | ``C`1``
+         * `Base<C<T3>>` | `"0.0.1"` | `int`
+         */
+        pragma[nomagic]
+        predicate hasBaseTypeMention(
+          Access a, AccessPosition apos, TypeMention baseMention, TypePath path, Type t
+        ) {
+          exists(Type sub | sub = inferRootType(a, apos) |
+            baseTypeMentionHasNonTypeParameterAt(sub, baseMention, path, t)
+            or
+            exists(TypePath prefix, TypePath suffix, TypeParameter tp |
+              baseTypeMentionHasTypeParameterAt(sub, baseMention, prefix, tp) and
+              t = inferTypeAt(a, apos, tp, suffix) and
+              path = prefix.append(suffix)
+            )
+          )
+        }
+      }
+
+      pragma[nomagic]
+      private predicate accessBaseType(
+        Access a, AccessPosition apos, Declaration target, Type base, TypePath path, Type t
+      ) {
+        exists(TypeMention tm |
+          target = a.getTarget() and
+          AccessBaseType::hasBaseTypeMention(a, apos, tm, path, t) and
+          base = resolveTypeMentionRoot(tm)
+        )
+      }
+
+      pragma[nomagic]
+      private predicate declarationBaseType(
+        Declaration decl, DeclarationPosition dpos, Type base, TypePath path, Type t
+      ) {
+        t = decl.getDeclaredType(dpos, path) and
+        path.isCons(base.getATypeParameter(), _)
+      }
+
+      /**
+       * Holds if the (transitive) base type `t` at `path` of `a` matches the type
+       * parameter `tp`, which is used in the declared types of `target`.
+       *
+       * For example, in
+       *
+       * ```csharp
+       * class C<T1> { }
+       *
+       * class Base<T2> {
+       * //         ^^ `tp`
+       *     public C<T2> Method() { ... }
+       * //               ^^^^^^ `target`
+       * }
+       *
+       * class Mid<T3> : Base<C<T3>> { }
+       *
+       * class Sub<T4> : Mid<C<T4>> { }
+       *
+       *    new Sub<int>().Method();
+       * // ^^^^^^^^^^^^^^^^^^^^^^^ `a`
+       * ```
+       *
+       * we have that type parameter `T2` of `Base` is matched as follows:
+       *
+       * `path`    | `t`
+       * --------- | -------
+       * `"0"`     | ``C`1``
+       * `"0.0"`   | ``C`1``
+       * `"0.0.1"` | `int`
+       */
+      pragma[nomagic]
+      private predicate baseTypeMatch(
+        Access a, Declaration target, TypePath path, Type t, TypeParameter tp
+      ) {
+        exists(AccessPosition apos, DeclarationPosition dpos, Type base, TypePath pathToTypeParam |
+          accessBaseType(a, apos, target, base, pathToTypeParam.append(path), t) and
+          declarationBaseType(target, dpos, base, pathToTypeParam, tp) and
```

Comment:
To check my understanding: The join of `base` here is the point where we introduce covariance, correct? Perhaps leave a comment in this spot about covariance if this is the place that needs updating to potentially support contravariance in the future?

Reply:
Correct.